### PR TITLE
Adds TagNode.fetch_or_create_by_xpath

### PR DESCRIPTION
**Pull request by funkyfuture**
It's surprisingly simple to implement an XPath parser. :grimacing:

### CODE REVIEW

File: _delb/nodes.py [hunk 1]
```diff
@@ -1703,6 +1704,101 @@ def document(self) -> Optional["Document"]:
...
+
+        All expressions must start with a ``./``.
+
+        >>> document = Document("<root/>")
+        >>> str(document.root.fetch_or_create_by_xpath("./child[@a='b']/grandchild"))
+        '<grandchild/>'
+        >>> str(document)
+        '<root><child a="b"><grandchild/></child></root>'
+        >>> str(document.root.fetch_or_create_by_xpath("./child[@a='b']/grandchild"))
+        '<grandchild/>'
```
Comment (funkyfuture) [inline _delb/nodes.py:1726]:
```suggestion
        >>> grandchild = document.root.fetch_or_create_by_xpath("./child[@a='b']/grandchild")
        >>> document.root.fetch_or_create_by_xpath("./child[@a='b']/grandchild") is grandchild
        True
```

**Comment by funkyfuture** [comment]
@JKatzwinkel that covers your cases, doesn't? is the method's docstring sufficiently comprehendible?

**Comment by JKatzwinkel** [comment]
Looks really great, method docstring makes sense, I will give it a try later, thanks!

File: _delb/xpath.py [hunk 2]
```diff
@@ -36,13 +38,19 @@
 
 def _split(expression: str, separator: str) -> Iterator[str]:
     assert separator not in ('"', "'")
+    cursor = -1
```
Comment (JKatzwinkel) [inline _delb/xpath.py:41]: I'm pretty sure `cursor` must be initialized with `0`.

File: _delb/xpath.py [hunk 3]
```diff
@@ -51,19 +59,58 @@ def _split(expression: str, separator: str) -> Iterator[str]:
...
+            return False
+
+        for step in steps[1:]:
+            if step.axis != "child":
+                return False
+
+            predicates = step.predicates
+            if not predicates:
+                continue
+            if len(predicates) > 1:
```
Comment (JKatzwinkel) [inline _delb/xpath.py:93]: I don't get why you wouldn't allow this, even though it's implemented, and concatenating predicates with `and` is explicitly permitted..?

File: _delb/xpath.py [hunk 4]
```diff
@@ -130,8 +181,39 @@ def __str__(self):
...
+    @lru_cache(1)
+    def parse_attribute_test_to_dict(self) -> Dict[str, Optional[str]]:
+        # assumes that this is called after the containing
+        # XPathExpression.is_unambiguously_locatable has returned True
+        if not self.type == "attribute_test":
+            raise ValueError
+
+        result: Dict[str, Optional[str]] = {}
+
+        for test in _split(self.expression, " and "):
```
Comment (JKatzwinkel) [inline _delb/xpath.py:209]: this would fail, due to a aforementioned tiny error in `_split`.

**Review by JKatzwinkel — CHANGES_REQUESTED** [review]
looks really beautiful, I'm gonna file a PR to alleviate that.

File: _delb/xpath.py (hunk 2 above)
Comment (funkyfuture) [reply _delb/xpath.py:41]: yep, likely. iirc i had changed the comparison operator in the loop and not adjusted this bit.

File: _delb/xpath.py (hunk 3 above)
Comment (funkyfuture) [reply _delb/xpath.py:93]: i was and am very unsure that/how/which multiple predicates can be grouped in square brackets, b/c i find the [spec](https://www.w3.org/TR/1999/REC-xpath-19991116/#predicates) very unspecific regarding a definition and it gives only examples where one of the predicates tests for a node's index. the latter is ignored by the implementation b/c there's no use for it.

also, the data model is inconsistent with regards to predicates, it follows the simple parsing not the grammar, see [this](https://github.com/funkyfuture/delb/blob/main/_delb/xpath.py#L90) and [that](https://github.com/funkyfuture/delb/pull/20/commits/83f36cc9333db82cbcca3f9077b0dd65bc7ff1e5#diff-4cebee66d5fda6b56b40455efd0a93eba562dfc41c1b3f27ee8d678e29bff537R16).

Comment (funkyfuture) [reply _delb/xpath.py:93]: okay, definition `4` [here](https://www.w3.org/TR/1999/REC-xpath-19991116/#predicates#section-Location-Steps) clears that up for me.

Comment (JKatzwinkel) [reply _delb/xpath.py:93]: do you mean `Step ::= | AxisSpecifier NodeTest Predicate*`, and does that indicate you can concatenate as many individually-bracketed predicates as you want?

File: _delb/xpath.py [hunk 5]
```diff
@@ -13,9 +13,11 @@
 # You should have received a copy of the GNU Affero General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+""" This is not an XPath implementation. """
```
Comment (JKatzwinkel) [inline _delb/xpath.py:16]: ah ok I thought this was a joke at first.

File: _delb/xpath.py (hunk 3 above)
Comment (funkyfuture) [reply _delb/xpath.py:93]: yes, and yes that's how i'd interpret it. i couldn't find BNFs for / definitions of possible predicates yet, which is one reason why i'm hesistating to do it more properly.

File: _delb/xpath.py (hunk 5 above)
Comment (funkyfuture) [reply _delb/xpath.py:16]: let's see where this ends up. it wouldn't be too stupid to aim more correctness b/c at some point an XPath parser and evaluator needs to be implemented in Rust. it's stupid for many other reasons though.

**Comment by funkyfuture** [comment]
thanks for the additional tests.

my more "correct" (less hackish, really) implementation of the predicates now only fails with expressions like `./cit[@type="translation"][@lang="en"]` that are first transformed to `./entry/sense/cit[(@type="translation") and (@lang="en")]`, but bracket parsing isn't implemented. currently i'd say that this doesn't need to be part of a future, delb-native implementation of a subset of the XPath-grammar. hence that initial normalization could yield a simpler expression. but then, there's the risk that someone really throws in something like `cit[@foo or @bar][@peng="pow"]`. and that then incorrectly transformed expression trickles through to lxml's xpath evaluator.

i already tend to a Verbot of overly complicated XPath expressions for now. and maybe one should aim for a first spec of a supported subset of XPath in the `0.4` release? @03b8, @JKatzwinkel, @zed-g, do you have any thoughts on that?

**Comment by funkyfuture** [comment]
while elaborating the issues with @JKatzwinkel we concluded:

- round brackets must be parsed in principle
- probably though expressions with round brackets will raise an exception for now, to get forward towards the `0.3` release
- while multiple expressions in square brackets are (beside the special case as index test) one way too much to express `(EXPR) and (EXPR)…`, its handling is simple enough to keep it supported

oh, and btw, here's an attempt to implement XPath with a generic parser: https://github.com/emory-libraries/eulxml/tree/master/eulxml/xpath

**Comment by funkyfuture** [comment]
it certainly could use more tests, but parenthesized expressions should now be parsed properly.

but yes, functions now pose an unsolved problem. me currently thinks that only functions where an attribute makes sense as first argument and that return a boolean are simple enough to be supported. that are `starts-with` and `contains`. and `not` makes sense.

**Comment by JKatzwinkel** [comment]
While adding arbitrary assertions to [`test_quotes_in_css_selector`](https://github.com/funkyfuture/delb/blob/d9fd3d83264cbb1e68caf5a2608f9cea644b5ab4/tests/test_queries.py#L164-L170), I realized that any specification of an xpath subset we end up supporting would require us to *also* specify what parts of css we do and don't support... For instance, at this point, even after "implementation" of boolean xpath functions "support", we do not support the css selector `[attribute$=value]`, because the css library in use doesn't convert `$=` into xpath's `ends-with()` function, but rather into an expression involving `substring()`, the return type of which isn't boolean.

File: _delb/xpath.py [hunk 6]
```diff
@@ -74,6 +79,10 @@ def _partition_terms(expression: str) -> List[str]:  # noqa: C901
             elif character == "(":
                 bracket_level += 1
             elif character == ")":
+                if bracket_level == function_level and any(
+                    current_term.startswith(func) for func in BOOLEAN_FUNCTIONS
```
Comment (funkyfuture) [inline _delb/xpath.py:83]: neat, that's much more clever than my idea of looking ahead.

File: _delb/xpath.py [hunk 7]
```diff
@@ -83,6 +92,9 @@ def _partition_terms(expression: str) -> List[str]:  # noqa: C901
             current_term += character
         elif character == "(":
             bracket_level += 1
+            if current_term in BOOLEAN_FUNCTIONS:
```
Comment (funkyfuture) [inline _delb/xpath.py:95]: i mean this one actually.

File: _delb/xpath.py (hunk 6 above)
Comment (JKatzwinkel) [reply _delb/xpath.py:83]: thanks, obviously i considered lookahead at first as well, before realizing that with everything you already had put in place we could just look behind and be done with it `:3`

Comment (JKatzwinkel) [reply _delb/xpath.py:83]: i'm not sure as to what consequences might arise from me having forgotten to reset `function_level` after exiting a function expression...

File: _delb/xpath.py (hunk 7 above)
Comment (JKatzwinkel) [reply _delb/xpath.py:95]: yeah i'm pretty proud of that one ngl

File: _delb/xpath.py [hunk 8]
```diff
@@ -43,7 +43,7 @@
 )
 # ordered by weight (ascending)
 BOOLEAN_OPERATORS = ("or", "and", "!=", "=", ">", ">=", "<", "<=")
-BOOLEAN_FUNCTIONS = ("not", "starts-with", "contains")  # "substring"
+BOOLEAN_FUNCTIONS = ("contains", "not", "starts-with", "substring")
```
Comment (funkyfuture) [inline _delb/xpath.py:46]:
```suggestion
BOOLEAN_FUNCTIONS = ("contains", "not", "starts-with")
```

**Comment by funkyfuture** [comment]
there's no 'native' `ends-with` function in the XPath 1.0 specs, but it should be specified b/c of consistency w/ `starts-with`.

File: _delb/xpath.py [hunk 9]
```diff
@@ -129,9 +311,248 @@ def __str__(self):
...
+        self.value = value
+
+    @staticmethod
+    # related: https://github.com/python/mypy/issues/5107
+    def parse(expression: str) -> "AttributePredicate":  # type: ignore
+        """
+        Parse an xpath predicate string expression into an ``AttributePredicate``
+        instance.
+
+        >>> type(AttributePredicate.parse('@type="translation"')) == AttributePredicate
```
Comment (funkyfuture) [inline _delb/xpath.py:426]:
```suggestion
        >>> isinstance(AttributePredicate.parse('@type="translation"'), AttributePredicate)
```

File: CHANGES.rst [hunk 10]
```diff
@@ -16,8 +16,13 @@ Intentionally.
...
 
-- Subclasses can claim to be the default class based on the evaluation of a
-  document's content and configuration by implementing `__class_test__`.
+- Adds the :meth:`delb.TagNode.fetch_or_create_by_xpath` method.
+    - ⚠️ Because of that a pre-mature parser of XPath expressions has been
+      implemented and you can expect some expressions to cause failures, e.g.
+      with functions that take more than one argument.
+- Subclasses of :class:`delb.Document` can claim to be the default class based
+  on the evaluation of a document's content and configuration by implementing
+  `__class_test__`.
```
Comment (funkyfuture) [inline CHANGES.rst:25]:
```suggestion
  ``__class_test__``.
```

**Comment by funkyfuture** [comment]
thanks a lot for the valuable input. i wanted to put an end to this PR before it grew endlessly.

**Pull request merged**